src/lem/circuit.rs

```diff
@@ -0,0 +1,927 @@
+                        )
+                        .with_context(|| "failed to constrain `and`")?;
+
+                        concrete_path_vec.push(allocated_has_match);
+
+                        let saved_slot = &mut slots_count.clone();
+                        recurse(
+                            &mut cs.namespace(|| format!("{}", tag)),
+                            op,
+                            concrete_path_and_has_match,
+                            saved_slot,
+                            g,
+                        )?;
+                    }
+
+                    // Now we need to enforce that at exactly one path was taken. We do that by enforcing
+                    // that the sum of the previously collected `Boolean`s is one. But, of course, this
+                    // irrelevant if we're on a virtual path and thus we use an implication gadget.
+                    enforce_selector_with_premise(
+                        &mut cs.namespace(|| "enforce_selector_with_premise"),
+                        &concrete_path,
+                        &concrete_path_vec,
+                    )
+                    .with_context(|| " couldn't constrain `enforce_selector_with_premise`")?;
+                    Ok(())
+                }
+                LEMCTL::MatchSymbol(..) => Ok(()),
+                LEMCTL::Seq(ops, rest) => {
+                    for op in ops {
+                        macro_rules! constrain_slot {
+                            ( $preimg: expr, $img: expr, $allocated_preimg: expr, $allocated_img: expr, $slot: expr ) => {
+                                // Retrieve the preallocated preimage and image for this slot
+                                let (preallocated_preimg, preallocated_img) =
+                                    match $slot {
+                                        SlotType::Hash2 => &g.preallocated_hash2_slots[slots_count.consume_hash2()],
+                                        SlotType::Hash3 => &g.preallocated_hash3_slots[slots_count.consume_hash3()],
+                                        SlotType::Hash4 => &g.preallocated_hash4_slots[slots_count.consume_hash4()],
+                                    };
+
+                                // Add the implication constraint for the image
+                                implies_equal(
+                                    &mut cs.namespace(|| {
+                                        format!("implies equal for {}'s hash (LEMOP {:?})", $img, &op)
+                                    }),
+                                    &concrete_path,
+                                    $allocated_img.hash(),
+                                    &preallocated_img,
+                                )?;
+
+                                // For each component of the preimage, add implication constraints
+                                // for its tag and hash
+                                for (i, allocated_ptr) in $allocated_preimg.iter().enumerate() {
```

Comment:
I think there are redundant constraints here, related to the duplicate allocations when hashing and unhashing. You are adding implication constraints for every image and preimage element on every use of the slot, but in reality, only the image or preimage needs to be constrained on any given usage — and the other allocated element(s) can be returned directly.